posix: semaphore: fix bugs and simplify code

jmtorrespalma opened the pull request:
Modifies several functions that are causing wrong
behaviour.

 * semaphore.h: add missing restrict keyword.
 * sem_destroy(): check that nobody is waiting
   before destroying the object.
 * sem_timedwait(): simpify function logic and
   fix a bug when abstime > currtime, that passed
   ticks instead of ms to k_sem_take().
 * sem_wait(): avoid unnecessary checks.

Signed-off-by: Juan Manuel Torres Palma <j.m.torrespalma@gmail.com>

jmtorrespalma committed:
posix: semaphore: fix bugs and simplify code

Modifies several functions that are causing wrong
behaviour.

 * semaphore.h: add missing restrict keyword.
 * sem_destroy(): check that nobody is waiting
   before destroying the object.
 * sem_timedwait(): simpify function logic and
   fix a bug when abstime > currtime, that passed
   ticks instead of ms to k_sem_take().
 * sem_wait(): avoid unnecessary checks.
 * sem_init(): add pshared value assertion.

Signed-off-by: Juan Manuel Torres Palma <j.m.torrespalma@gmail.com>

lpereira commented on kernel/posix/semaphore.c:
```diff
@@ -51,6 +53,9 @@ int sem_init(sem_t *semaphore, int pshared, unsigned int value)
+	/*
+	 * XXX: this is not in the standard
+	 */
 	if (pshared != 0) {
```
This can be changed to just an assertion -- setting `ENOSYS` here is indeed wrong.

jmtorrespalma replied:
I agree with you there too, just wanted to point it out with a comment to get some feedback on it.

lpereira reviewed (commented):
Thanks for looking at this! Patch looks fine, I would just go ahead and move the pshared check to an assertion as it's not really standard.

lpereira commented on kernel/posix/semaphore.c:
```diff
@@ -115,17 +99,21 @@ int sem_timedwait(sem_t *semaphore, struct timespec *abstime)
 	}
 
-	return take_and_convert(semaphore, timeout);
+	if (k_sem_take(semaphore, timeout)) {
```
Shouldn't errno be set to EAGAIN if this timeouts?

jmtorrespalma committed:
tests: posix: fix semaphore.h test

Previous implementation was buggy, destroying
resources with pending thread and race conditions.

New version tests basic functionality with 2
threads and basic synchronization.

Signed-off-by: Juan Manuel Torres Palma <j.m.torrespalma@gmail.com>

jmtorrespalma commented:
Addressed assertion in `sem_init()`, while fixing several typos in docs.
Also new commit to fix test program, that I couldn't really understand the logic behind it. New version is very simple but tests basic functionality.
